File: tools/Word Document Editor/scripts/add_change.py
#!/usr/bin/env python3
"""
Add a new tracked change to a Word document.

Usage:
    python3 add_change.py document.docx --text "New text" --type insertion
    python3 add_change.py document.docx --text "Old text" --type deletion --position 100
    python3 add_change.py document.docx --text "New text" --type insertion --author "Custom Author"
"""

import sys
import zipfile
import shutil
import argparse
from xml.etree import ElementTree as ET
from pathlib import Path
from docx_utils import NS


def add_tracked_change(docx_path, text, change_type, author="AIM", position=None, output_path=None):
    """
    Add a new tracked change to a Word document.
    
    Args:
        docx_path: Path to input document
        text: Text content for the change
        change_type: 'insertion' or 'deletion'
        author: Author name for the change (default: "AIM")
        position: Character position (None = end of document)
        output_path: Path to save output (None = overwrites original)
    """
    docx_path = Path(docx_path)
    
    if not docx_path.exists():
        print(f"❌ Error: File not found: {docx_path}", file=sys.stderr)
        sys.exit(1)
    
    if change_type not in ['insertion', 'deletion']:
        print(f"❌ Error: Change type must be 'insertion' or 'deletion'", file=sys.stderr)
        sys.exit(1)
    
    # Determine output path
    if output_path is None:
        output_path = docx_path
    else:
        output_path = Path(output_path)
    
    try:
        # Copy document to output location
        if output_path != docx_path:
            shutil.copy2(docx_path, output_path)
        
        # Open document
        with zipfile.ZipFile(output_path, 'a') as docx:
            # Read document.xml
            doc_xml = docx.read('word/document.xml')
            root = ET.fromstring(doc_xml)
            
            # Get current date/time for the change
            from datetime import datetime
            change_date = datetime.now().isoformat() + 'Z'
            
            if change_type == 'insertion':
                # Create insertion element
                ins_elem = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}ins')
                ins_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', author)
                ins_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', change_date)
                ins_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id', '0')
                
                # Create run element with text
                run_elem = ET.SubElement(ins_elem, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}r')
                text_elem = ET.SubElement(run_elem, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}t')
                text_elem.text = text
                
                # Find insertion point (simplified - appends to end of body)
                body = root.find('.//w:body', NS)
                if body is not None:
                    # Find last paragraph or create one
                    paragraphs = body.findall('.//w:p', NS)
                    if paragraphs:
                        last_para = paragraphs[-1]
                        # Add to end of last paragraph
                        last_para.append(ins_elem)
                    else:
                        # Create new paragraph
                        para = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}p')
                        para.append(ins_elem)
                        body.append(para)
            
            elif change_type == 'deletion':
                # Create deletion element
                del_elem = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}del')
                del_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', author)
                del_elem.set('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', change_date)
                
                # Create run element with text
                run_elem = ET.SubElement(del_elem, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}r')
                text_elem = ET.SubElement(run_elem, '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}delText')
                text_elem.text = text
                
                # Find insertion point (simplified - appends to end of body)
                body = root.find('.//w:body', NS)
                if body is not None:
                    paragraphs = body.findall('.//w:p', NS)
                    if paragraphs:
                        last_para = paragraphs[-1]
                        last_para.append(del_elem)
                    else:
                        para = ET.Element('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}p')
                        para.append(del_elem)
                        body.append(para)
            
            # Write back to document
            doc_xml_new = ET.tostring(root, encoding='unicode')
            # Word requires specific XML declaration
            doc_xml_new = '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>' + doc_xml_new
            
            # Remove old file and write new one
            docx.writestr('word/document.xml', doc_xml_new.encode('utf-8'))
        
        print(f"\n✓ Success! Tracked change added and saved to: {output_path}")
        print(f"   Type: {change_type}")
        print(f"   Author: {author}")
        print(f"   Text: {repr(text)}")
        
    except Exception as e:
        print(f"❌ Error: {type(e).__name__}: {str(e)}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(
        description='Add a new tracked change to a Word document'
    )
    parser.add_argument('document', help='Path to Word document (.docx)')
    parser.add_argument('--text', required=True,
                       help='Text content for the change')
    parser.add_argument('--author', default='AIM',
                       help='Author name for the change (default: AIM)')
    parser.add_argument('--type', required=True, choices=['insertion', 'deletion'],
                       help='Type of change: insertion or deletion')
    parser.add_argument('--position', type=int,
                       help='Character position (optional, defaults to end)')
    parser.add_argument('--output', '-o',
                       help='Output file path (default: overwrites original)')
    
    args = parser.parse_args()
    
    add_tracked_change(
        args.document,
        args.text,
        args.type,
        author=args.author,
        position=args.position,
        output_path=args.output
    )


if __name__ == '__main__':
    main()

File: tools/Word Document Editor/scripts/add-change.js
#!/usr/bin/env node
/**
 * Add a new tracked change to a Word document.
 *
 * Usage:
 *     node add-change.js document.docx --text "New content" --type insertion
 */

import { readDocx, writeDocx, parseXml, serializeXml, select } from './docx-utils.js';
import { existsSync, copyFileSync } from 'fs';

/**
 * Add a tracked change to a Word document
 */
async function addChange(docxPath, text, options = {}) {
  const { type = 'insertion', author = 'AIM', outputPath = null } = options;

  // Determine output path
  const finalOutputPath = outputPath || docxPath;

  // Copy file if output is different
  if (outputPath && outputPath !== docxPath) {
    copyFileSync(docxPath, outputPath);
  }

  // Read document
  const zip = await readDocx(outputPath || docxPath);

  // Read document.xml
  const docXml = await zip.file('word/document.xml')?.async('string');
  if (!docXml) {
    throw new Error('Invalid Word document structure');
  }
  const docDoc = parseXml(docXml);

  const changeDate = new Date().toISOString();
  const nsW = 'http://schemas.openxmlformats.org/wordprocessingml/2006/main';

  // Find body
  const body = select('//w:body', docDoc)[0];
  if (!body) {
    throw new Error('Document body not found');
  }

  // Create the change element based on type
  let changeElem;
  if (type === 'insertion') {
    changeElem = docDoc.createElementNS(nsW, 'w:ins');
  } else if (type === 'deletion') {
    changeElem = docDoc.createElementNS(nsW, 'w:del');
  } else {
    throw new Error(`Unsupported change type: ${type}. Use 'insertion' or 'deletion'.`);
  }

  changeElem.setAttribute('w:author', author);
  changeElem.setAttribute('w:date', changeDate);

  // Create paragraph with run and text
  const para = docDoc.createElementNS(nsW, 'w:p');
  const run = docDoc.createElementNS(nsW, 'w:r');
  const textElem = docDoc.createElementNS(nsW, 'w:t');
  textElem.textContent = text;

  run.appendChild(textElem);

  if (type === 'deletion') {
    // For deletions, use w:delText instead of w:t
    const delTextElem = docDoc.createElementNS(nsW, 'w:delText');
    delTextElem.textContent = text;
    run.removeChild(textElem);
    run.appendChild(delTextElem);
  }

  changeElem.appendChild(run);
  para.appendChild(changeElem);

  // Append to body
  body.appendChild(para);

  // Write back document.xml
  zip.file('word/document.xml', serializeXml(docDoc));

  // Enable track changes in settings.xml if it exists
  const settingsXml = await zip.file('word/settings.xml')?.async('string');
  if (settingsXml && !settingsXml.includes('trackRevisions')) {
    const newSettings = settingsXml.replace(
      '</w:settings>',
      '  <w:trackRevisions/>\n</w:settings>'
    );
    zip.file('word/settings.xml', newSettings);
  }

  // Save
  await writeDocx(zip, finalOutputPath);

  return { type, author, text };
}

// CLI
function showHelp() {
  console.log(`
Add a new tracked change to a Word document

Usage:
  node add-change.js <document> --text "Text" --type <type> [options]

Arguments:
  document        Path to Word document (.docx)

Options:
  --text TEXT     Text to insert or delete (required)
  --type TYPE     Type: 'insertion' or 'deletion' (default: insertion)
  --author NAME   Author name for the change (default: AIM)
  --output, -o    Output file path (default: overwrites original)
  --help, -h      Show this help message

Examples:
  node add-change.js document.docx --text "New content" --type insertion
  node add-change.js document.docx --text "Removed text" --type deletion
  node add-change.js document.docx --text "Edit" --author "Reviewer" --output modified.docx
`);
  process.exit(0);
}

const args = process.argv.slice(2);

if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
  showHelp();
}

const docxPath = args[0];
let text = null;
let type = 'insertion';
let author = 'AIM';
let outputPath = null;

for (let i = 1; i < args.length; i++) {
  if (args[i] === '--text' && args[i + 1]) {
    text = args[++i];
  } else if (args[i] === '--type' && args[i + 1]) {
    type = args[++i];
  } else if (args[i] === '--author' && args[i + 1]) {
    author = args[++i];
  } else if ((args[i] === '--output' || args[i] === '-o') && args[i + 1]) {
    outputPath = args[++i];
  }
}

if (!text) {
  console.error('Error: --text is required');
  process.exit(1);
}

if (!existsSync(docxPath)) {
  console.error(`Error: File not found: ${docxPath}`);
  process.exit(1);
}

try {
  const result = await addChange(docxPath, text, { type, author, outputPath });
  console.log(`\nSuccess! Change added and saved to: ${outputPath || docxPath}`);
  console.log(`   Type: ${result.type}`);
  console.log(`   Author: ${result.author}`);
  console.log(`   Text: "${result.text}"`);
} catch (e) {
  console.error(`Error: ${e.message}`);
  process.exit(1);
}

